sympy/functions/combinatorial/factorials.py

```diff
@@ -1057,3 +1055,96 @@ def _eval_is_nonnegative(self):
                 return True
             elif k.is_even is False:
                 return  False
+
+class multinomial(CombinatorialFunction):
+    r"""Implementation of the multinomial coefficient. They can be expressed as
+    a product of binomial coefficients or of factorials:
+
+    .. math:: \frac{(k_1+k_2+\cdots+k_m)!}{k_1!\, k_2! \cdots k_m!} = {k_1\choose k_1}{k_1+k_2\choose k_2}\cdots{k_1+k_2+\cdots+k_m\choose k_m}
+
+    The multinomial coefficients have a direct combinatorial interpretation,
+    as the number of ways of depositing n distinct objects into m distinct bins,
+    with k1 objects in the first bin, k2 objects in the second bin, and so on.
+
+    To expand the multinomial when `k` contains a symbol, use either
+    ``expand_func()`` or ``expand(func=True)``. The former will keep
+    the polynomial in factored form while the latter will expand the
+    polynomial itself. See examples for details.
+
+    Examples
+    ========
+
+    >>> from sympy import Symbol, Rational, multinomial, expand_func, symbols
+    >>> n = Symbol('n', integer=True, positive=True)
+
+    >>> multinomial(15, 8)
+    490314
+
+    >>> multinomial(n, -1)
+    0
+
+    Comparing Coefficents generated with the multinomial function:
+
+    >>> a, b, c = symbols('a b c')
+    >>> ((a+b+c)**3).expand(func=True)
+    a**3 + 3*a**2*b + 3*a**2*c + 3*a*b**2 + 6*a*b*c + 3*a*c**2 + b**3 + 3*b**2*c + 3*b*c**2 + c**3
+
+    The Coefficient when of 2 1 0 corresponding to the second term is 3
+
+    >>> multinomial(2,1,0)
+    3
+
+    >>> multinomial(Rational(5, 4), 3)
+    gamma(21/4)/(6*gamma(9/4))
+
+    >>> multinomial(n, 3)
+    gamma(n + 4)/(6*gamma(n + 1))
+
+    >>> multinomial(n, 3).expand(func=True)
+    n**3/6 + n**2 + 11*n/6 + 1
+
+    >>> expand_func(multinomial(n, 3))
+    (n + 1)*(n + 2)*(n + 3)/6
+
+    References
+    ==========
+
+    .. [1] https://en.wikipedia.org/wiki/Multinomial_theorem#Multinomial_coefficients
+
+    .. [2] http://functions.wolfram.com/GammaBetaErf/Multinomial/02/0001/
+
+    """
+
+    @classmethod
+    def eval(self, *args):
+        args = list(map(sympify, args))
+        if all(x.is_Integer for x in args):
```

Comment:
it seems like you could just handle everything with the first if:
```python
def runsum(args):
    t = 0
    for i in args:
        t += i
        yield t

>>> k=(1,2,3,pi,x)
>>> Mul(*[binomial(j,k) for j,k in zip(runsum(k),k)])
binomial(x + pi + 6, x)*gamma(pi + 7)/(12*gamma(1 + pi))
```

Reply:
This raises the question of when it should remain unevaluated...only for all symbolic args? Should everything up to symbols be evaluated completely?

Reply:
> it seems like you could just handle everything with the first if:

well, you are mostly correct. The thing is Multinomial is not just limited to integers. It takes in non-integral values as well. Since the formal definition of factorials sticks to integers, gamma is preferred when it comes to non-integral and complex numbers for which `_.is_number ` tends to evaluate to `True`.
Though factorial in Sympy supports non-integers,
I mean check out the difference 
```
>>> factorial(S.Half)
factorial(1/2)
>>> factorial(1/2)
0.886226925452758
>>> gamma(S.Half+1)
sqrt(pi)/2
```
anddd this is the method I've implemented.

```
>>> k=(3,pi,symbols('x'))
>>> multinomial(*k)
multinomial(3, pi, x)
>>> multinomial(*k).rewrite(gamma)
gamma(x + pi + 4)/(6*gamma(1 + pi)*gamma(x + 1))
>>> multinomial(*k).rewrite(factorial)
factorial(x + 3 + pi)/(6*factorial(pi)*factorial(x))
>>> multinomial(*k).rewrite(binomial)
binomial(x + 3 + pi, x)*gamma(pi + 4)/(6*gamma(1 + pi))
>>> multinomial(*k).expand(func=True)
x**3*gamma(x + pi)/(6*pi*gamma(pi)*gamma(x)) + x**2*gamma(x + pi)/(pi*gamma(pi)*gamma(x)) + 2*x**2*gamma(x + pi)/(3*gamma(pi)*gamma(x)) + 11*x*gamma(x + pi)/(6*pi*gamma(pi)*gamma(x)) + 3*x*gamma(x + pi)/(gamma(pi)*gamma(x)) + pi*x*gamma(x + pi)/(gamma(pi)*gamma(x)) + gamma(x + pi)/(pi*gamma(pi)*gamma(x)) + 11*gamma(x + pi)/(3*gamma(pi)*gamma(x)) + 2*pi**2*gamma(x + pi)/(3*gamma(pi)*gamma(x)) + 3*pi*gamma(x + pi)/(gamma(pi)*gamma(x)) + gamma(x + pi)/(x*gamma(pi)*gamma(x)) + pi**3*gamm
a(x + pi)/(6*x*gamma(pi)*gamma(x)) + 11*pi*gamma(x + pi)/(6*x*gamma(pi)*gamma(x)) + pi**2*gamma(x + pi)/(x*gamma(pi)*gamma
(x))
>>> expand_func(multinomial(*k))
(x + pi)*(x + 1 + pi)*(x + 2 + pi)*(x + 3 + pi)*gamma(x + pi)/(6*pi*x*gamma(pi)*gamma(x))
```
this tends to work for all cases, only because it is left unevaluated if symbols are present. I've written it such that it remains unevaluated unless all parameters are numbers.
```
>>> multinomial(n,n,n)
multinomial(n, n, n)
>>> multinomial(1,n,n)
multinomial(1, n, n)
>>> multinomial(1,2,n)
multinomial(1, 2, n)
```
There is also an issue with the binomial expansion, one I intend to raise as an issue and correct in the future. But multinomial doesn't exhibit the same erroneous behaviour.
```
>>> binomial(n,n)
binomial(n, n)
>>> binomial(n,n-3).expand(func=True)
n**3/6 - n**2/2 + n/3
>>> binomial(2*n,n-3).expand(func=True)
binomial(2*n, n - 3)
>>> expand_func(binomial(2*n,n-1))
binomial(2*n, n - 1)
```
```
>>> multinomial(n*2,n-1).expand(func=True)
gamma(3*n)/(2*n*gamma(n)*gamma(2*n))
```